contributors/design-proposals/gce-l4-loadbalancer-healthcheck.md

```diff
@@ -0,0 +1,57 @@
+# GCE L4 load-balancers' health checks for nodes
+
+## Goal
+Set up health checks for GCE L4 load-balancer to ensure it is only
+targeting healthy nodes.
+
+## Motivation
+On cloud providers which support external load balancers, setting the
+type field to "LoadBalancer" will provision a L4 load-balancer for the
+service ([doc](https://kubernetes.io/docs/concepts/services-networking/service/#type-loadbalancer)),
+which load-balances traffics to k8s nodes. As of k8s 1.6, we don't
+create health check for L4 load-balancer by default, which means all
+traffics will be forwarded to any one of the nodes blindly.
+
+This is undesired in cases:
+- k8s components including kubelet dead on nodes. Nodes will be flipped
+to unhealthy after a long propagation (~40s), even if we remove nodes
+from target instance group at that point it is too slow.
+- kube-proxy dead on nodes while kubelet is still alive. Requests will
+be continually forwarded to nodes that may not be able to properly route
+traffics.
+
+For now, the only case health check will be created is for
+[OnlyLocal Service](https://kubernetes.io/docs/tutorials/services/source-ip/#source-ip-for-services-with-typeloadbalancer).
+We should have a node-level health check for load balancers that are used
+by non-OnlyLocal services.
+
+## Design
+Healthchecking the kube-proxys seems to be the best choice:
+- kube-proxy runs on every nodes and it is the pivot for service traffic
+routing.
+- Port 10249 on nodes is currently used for both kube-proxy's healthz and
+pprof.
+- We already have a similar mechanism for healthchecking OnlyLocal services
+in kube-proxy.
+
+We should enable health check by default for non-OnlyLocal LoadBalancer
+service but provide a way for opting out initially.
+
+## Implementation
+
+API
+- Define an annotation on service for opting out load-balancer health check.
+Eventually this annotation will be deprecated and all load-balancers should
+come with health check.
+
+kube-proxy
+- Make healthz bind to 0.0.0.0 by default (for now we only bind localhost).
+As we only allow traffics from load-balancer source IPs to healthz port,
+this wouldn't be a big security concern.
+
+GCE cloud provider
+- Manage `k8s-l4-default-healthcheck` firewall and healthcheck resources.
```

Comment:
Nick told me offline, health check for L4 load-balancer currently comes from 169.254.169.254. In the future it will come from other external IPs so we will need additional firewall rule.

And the naming is a good point :)